File: internal/app/describe.go
package app

import (
	"encoding/json"
	"fmt"
	"io"

	"github.com/leandronsp/githerb/internal/review"
)

// Describe attaches the reasoning to a proposal: the decisions it carries and
// the lines the author wants to explain before anyone asks.
//
// It takes JSON rather than flags because the writer is usually an agent, and
// the domain caps every field, so a description cannot ramble no matter which
// agent or which harness produced it.
type Describe struct {
	Proposals review.Proposals
	Author    string
	Now       Clock
}

// Description is the shape an author submits.
type Description struct {
	Chunks    []ChunkInput     `json:"chunks"`
	Rationale []RationaleInput `json:"rationale"`
}

// ChunkInput is one reviewable decision. Every field is one line and the
// domain refuses anything longer than its ceiling.
type ChunkInput struct {
	Title    string `json:"title"`
	Surface  string `json:"surface"`
	Before   string `json:"before"`
	After    string `json:"after"`
	Decision string `json:"decision"`
	Rejected string `json:"rejected,omitempty"`
	File     string `json:"file,omitempty"`
	Side     string `json:"side,omitempty"`
	Start    int    `json:"start,omitempty"`
	End      int    `json:"end,omitempty"`
}

// RationaleInput is the author explaining a few lines.
type RationaleInput struct {
	File  string `json:"file"`
	Side  string `json:"side,omitempty"`
	Start int    `json:"start"`
	End   int    `json:"end,omitempty"`
	Body  string `json:"body"`
}

// Run reads the description and writes it against the head revision.
func (d Describe) Run(id string, from io.Reader) (int, error) {
	proposal, err := d.Proposals.Load(review.ProposalID(id))
	if err != nil {
		return 0, err
	}

	var description Description

	if err := json.NewDecoder(from).Decode(&description); err != nil {
		return 0, fmt.Errorf("reading the description: %w", err)
	}

	head := proposal.Head().SHA()
	written := 0

	for _, input := range description.Chunks {
		chunk, err := d.chunk(input)
		if err != nil {
			return written, err
		}

		if err := d.Proposals.Annotate(head, review.ChunkRecord(chunk)); err != nil {
			return written, err
		}

		written++
	}

	for _, input := range description.Rationale {
		comment, err := d.rationale(input, head)
		if err != nil {
			return written, err
		}

		if err := d.Proposals.Annotate(head, review.RationaleRecord(comment)); err != nil {
			return written, err
		}

		written++
	}

	return written, nil
}

func (d Describe) chunk(input ChunkInput) (review.Chunk, error) {
	chunk, err := review.NewChunk(
		input.Title, input.Surface, input.Before, input.After, input.Decision, input.Rejected,
	)
	if err != nil {
		return review.Chunk{}, err
	}

	if input.File == "" {
		return chunk, nil
	}

	span, err := spanOf(input.Side, input.Start, input.End)
	if err != nil {
		return review.Chunk{}, err
	}

	anchored, err := chunk.At(review.File(input.File), span)
	if err != nil {
		return review.Chunk{}, err
	}

	return anchored, nil
}

func (d Describe) rationale(input RationaleInput, head review.SHA) (review.Comment, error) {
	span, err := spanOf(input.Side, input.Start, input.End)
	if err != nil {
		return review.Comment{}, err
	}

	comment, err := review.NewComment(head, review.File(input.File), span, input.Body, d.Author, d.Now())
	if err != nil {
		return review.Comment{}, err
	}

	return comment, nil
}

// spanOf fills in what an author left out: the new side, and a single line.
func spanOf(side string, start, end int) (review.Span, error) {
	if side == "" {
		side = string(review.SideNew)
	}

	if end == 0 {
		end = start
	}

	parsed, err := review.ParseSide(side)
	if err != nil {
		return review.Span{}, err
	}

	span, err := review.NewSpan(parsed, start, end)
	if err != nil {
		return review.Span{}, err
	}

	return span, nil
}

File: crates/app/src/describe.rs
//! The reasoning, attached to a proposal: the decisions it carries and the
//! lines the author wants to explain before anybody asks.
//!
//! It takes JSON rather than flags because the writer is usually an agent,
//! and every field has a ceiling in the core, so a description cannot ramble
//! whichever agent or harness produced it.

use review::{
    Anchor, Author, Chunk, FilePath, ProposalId, Rationale, Record, Side, Span, Timestamp,
};
use serde::Deserialize;

use crate::error::{Error, Result};
use crate::store::Store;

/// The shape an author submits.
#[derive(Debug, Default, Deserialize)]
struct Description {
    #[serde(default)]
    chunks: Vec<ChunkInput>,
    #[serde(default)]
    rationale: Vec<RationaleInput>,
}

/// One reviewable decision, as it arrives.
#[derive(Debug, Default, Deserialize)]
struct ChunkInput {
    #[serde(default)]
    title: String,
    #[serde(default)]
    surface: String,
    #[serde(default)]
    before: String,
    #[serde(default)]
    after: String,
    #[serde(default)]
    decision: String,
    #[serde(default)]
    rejected: String,
    #[serde(default)]
    file: String,
    #[serde(default)]
    side: String,
    #[serde(default)]
    start: u32,
    #[serde(default)]
    end: u32,
}

/// The author explaining a few lines, as it arrives.
#[derive(Debug, Default, Deserialize)]
struct RationaleInput {
    #[serde(default)]
    file: String,
    #[serde(default)]
    side: String,
    #[serde(default)]
    start: u32,
    #[serde(default)]
    end: u32,
    #[serde(default)]
    body: String,
}

/// Read a description and write it against the head revision.
///
/// Chunks first, then rationale, in the order they were given. A refusal
/// halfway through leaves what came before it in the log, because the log is
/// append-only and half a description is still true.
///
/// # Errors
///
/// JSON this build cannot read, a field over its ceiling or carrying a
/// newline, a chunk missing its before, after or decision, or a span that
/// covers no line.
pub fn describe(
    store: &Store,
    author: &Author,
    now: Timestamp,
    id: &ProposalId,
    json: &str,
) -> Result<usize> {
    let description: Description =
        serde_json::from_str(json).map_err(|err| Error::Description(err.to_string()))?;

    let proposal = store.load(id)?;
    let head = proposal.head().sha().clone();
    let mut written = 0;

    for input in &description.chunks {
        store.annotate(&head, &Record::Chunk(chunk(input)?))?;
        written += 1;
    }

    for input in &description.rationale {
        let anchor = anchor(&input.file, &input.side, input.start, input.end)?;
        let note = Rationale::new(head.clone(), anchor, &input.body, author.clone(), now)?;

        store.annotate(&head, &Record::Rationale(note))?;
        written += 1;
    }

    Ok(written)
}

/// The shape a description takes, for somebody about to write one.
#[must_use]
pub fn template() -> &'static str {
    include_str!("describe/template.txt")
}

/// One chunk, anchored when it says where it is.
fn chunk(input: &ChunkInput) -> Result<Chunk> {
    let chunk = Chunk::new(
        &input.title,
        optional(&input.surface),
        &input.before,
        &input.after,
        &input.decision,
        optional(&input.rejected),
    )?;

    if input.file.trim().is_empty() {
        return Ok(chunk);
    }

    Ok(chunk.anchored(anchor(&input.file, &input.side, input.start, input.end)?))
}

/// Where a description points, filling in what the author left out: the new
/// side, and a single line.
fn anchor(file: &str, side: &str, start: u32, end: u32) -> Result<Anchor> {
    let side = if side.trim().is_empty() {
        Side::New
    } else {
        Side::parse(side)?
    };
    let end = if end == 0 { start } else { end };

    Ok(Anchor::new(
        FilePath::parse(file)?,
        Span::new(side, start, end)?,
    ))
}

/// A field an author may leave out entirely.
fn optional(value: &str) -> Option<&str> {
    Some(value).filter(|text| !text.trim().is_empty())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_span_defaults_to_one_line_on_the_new_side() -> Result<()> {
        let anchored = anchor("a.rs", "", 12, 0)?;

        assert_eq!(anchored.span().side(), Side::New);
        assert_eq!(anchored.span().start(), 12);
        assert_eq!(anchored.span().end(), 12);
        Ok(())
    }

    #[test]
    fn a_chunk_with_no_file_stays_unanchored() -> Result<()> {
        let input = ChunkInput {
            title: "the gate".to_owned(),
            before: "it was not there".to_owned(),
            after: "it is".to_owned(),
            decision: "put it in".to_owned(),
            ..ChunkInput::default()
        };

        assert_eq!(chunk(&input)?.anchor(), None);
        Ok(())
    }

    #[test]
    fn a_decision_over_its_ceiling_is_refused() {
        let input = ChunkInput {
            title: "the gate".to_owned(),
            before: "it was not there".to_owned(),
            after: "it is".to_owned(),
            decision: "x".repeat(201),
            ..ChunkInput::default()
        };

        let err = chunk(&input).unwrap_err();

        assert!(
            matches!(err, Error::Review(review::Error::TooLong { .. })),
            "{err}"
        );
    }

    #[test]
    fn the_template_is_the_shape_a_description_takes() {
        assert!(template().starts_with("{\n  \"chunks\": ["));
        assert!(template().ends_with("the good intentions of whoever wrote it.\n"));
    }
}
